threedworld-mit/tdw
# Added: send_field_of_view

# Test

1. In the `tdw` repo, `git fetch` and `git checkout send_field_of_view`
2. Download and extract [this test build](https://www.dropbox.com/s/2mp2lx14kh2gicx/TDW_Windows_send_field_of_view.zip?dl=0)
3. Run the test build.
4. Run this controller:

```python
from tdw.controller import Controller
from tdw.tdw_utils import TDWUtils
from tdw.output_data import OutputData, FieldOfView
from tdw.add_ons.third_person_camera import ThirdPersonCamera


c = Controller(launch_build=False)
camera = ThirdPersonCamera(avatar_id="a",
                           position={"x": 1, "y": 1, "z": 0},
                           look_at={"x": 0, "y": 0, "z": 0})
c.add_ons.append(camera)
c.communicate(TDWUtils.create_empty_room(12, 12))
resp = c.communicate([{"$type": "set_field_of_view",
                       "field_of_view": 55,
                       "avatar_id": "a"},
                      {"$type": "send_field_of_view",
                       "frequency": "always"}])
for i in range(len(resp) - 1):
    r_id = OutputData.get_data_type_id(resp[i])
    if r_id == "fofv":
        field_of_view = FieldOfView(resp[i])
        print(field_of_view.get_avatar_id())
        print(field_of_view.get_sensor_name())
        print(field_of_view.get_fov())
        print(field_of_view.get_focal_length())
c.communicate({"$type": "terminate"})
```

Expected output:

```
a
SensorContainer
55.0
34.57767868041992
```

# Changelog

### Command API

#### New Commands

| Command              | Description                                     |
| -------------------- | ----------------------------------------------- |
| `send_field_of_view` | Send the camera field of view and focal length. |

### Output Data

#### New Output Data

| Output Data   | Description                                |
| ------------- | ------------------------------------------ |
| `FieldOfView` | The camera field of view and focal length. |
